// @mongosh/shell-api/lib/shell-internal-state.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("@mongosh/errors");
const service_provider_core_1 = require("@mongosh/service-provider-core");
const events_1 = require("events");
const mongodb_redact_1 = __importDefault(require("mongodb-redact"));
const decorators_1 = require("./decorators");
const enums_1 = require("./enums");
const error_codes_1 = require("./error-codes");
const index_1 = require("./index");
const interruptor_1 = require("./interruptor");
const no_db_1 = __importDefault(require("./no-db"));
const shell_bson_1 = __importDefault(require("./shell-bson"));
class ShellInternalState {
    constructor(initialServiceProvider, messageBus = new events_1.EventEmitter(), cliOptions = {}) {
        var _a;
        this.batchSizeFromDBQuery = undefined;
        this.isInteractive = false;
        this.interrupted = new interruptor_1.InterruptFlag();
        this.initialServiceProvider = initialServiceProvider;
        this.messageBus = messageBus;
        this.shellApi = new index_1.ShellApi(this);
        this.shellBson = shell_bson_1.default(initialServiceProvider.bsonLibrary, (msg) => {
            if (this.context.print) {
                this.context.print(`Warning: ${msg}`);
            }
        });
        this.mongos = [];
        this.connectionInfo = { buildInfo: {} };
        if (!cliOptions.nodb) {
            const mongo = new index_1.Mongo(this, undefined, undefined, undefined, initialServiceProvider);
            this.mongos.push(mongo);
            this.currentDb = mongo.getDB(initialServiceProvider.initialDb || service_provider_core_1.DEFAULT_DB);
        }
        else {
            this.currentDb = new no_db_1.default();
        }
        this.currentCursor = null;
        this.context = {};
        this.cliOptions = cliOptions;
        this.evaluationListener = {};
        this.mongocryptdSpawnPath = (_a = cliOptions.mongocryptdSpawnPath) !== null && _a !== void 0 ? _a : null;
    }
    async fetchConnectionInfo() {
        if (!this.cliOptions.nodb) {
            this.connectionInfo = await this.currentDb._mongo._serviceProvider.getConnectionInfo();
            this.messageBus.emit('mongosh:connect', {
                ...this.connectionInfo.extraInfo,
                uri: mongodb_redact_1.default(this.connectionInfo.extraInfo.uri)
            });
        }
    }
    async close(force) {
        for (const mongo of [...this.mongos]) {
            await mongo.close(force);
        }
    }
    setDbFunc(newDb) {
        this.currentDb = newDb;
        this.context.rs = new index_1.ReplicaSet(this.currentDb);
        this.context.sh = new index_1.Shard(this.currentDb);
        this.fetchConnectionInfo().catch(err => this.messageBus.emit('mongosh:error', err));
        this.currentDb._getCollectionNamesForCompletion().catch(err => this.messageBus.emit('mongosh:error', err));
        this.currentDb._mongo._getDatabaseNamesForCompletion().catch(err => this.messageBus.emit('mongosh:error', err));
        return newDb;
    }
    setCtx(contextObject) {
        this.context = contextObject;
        Object.assign(contextObject, this.shellApi);
        for (const name of Object.getOwnPropertyNames(index_1.ShellApi.prototype)) {
            const { shellApi } = this;
            if (decorators_1.toIgnore.concat(['help']).includes(name) ||
                typeof shellApi[name] !== 'function') {
                continue;
            }
            contextObject[name] = function (...args) {
                return shellApi[name](...args);
            };
            contextObject[name].help = shellApi[name].help;
        }
        contextObject.help = this.shellApi.help;
        Object.assign(contextObject, this.shellBson);
        if (contextObject.console === undefined) {
            contextObject.console = {};
        }
        for (const key of ['log', 'warn', 'info', 'error']) {
            contextObject.console[key] = async (...args) => {
                return await contextObject.print(...args);
            };
        }
        contextObject.console.clear = contextObject.cls;
        contextObject.rs = new index_1.ReplicaSet(this.currentDb);
        contextObject.sh = new index_1.Shard(this.currentDb);
        const setFunc = (newDb) => {
            if (index_1.getShellApiType(newDb) !== 'Database') {
                throw new errors_1.MongoshInvalidInputError('Cannot reassign \'db\' to non-Database type', errors_1.CommonErrors.InvalidOperation);
            }
            return this.setDbFunc(newDb);
        };
        if (this.initialServiceProvider.platform === service_provider_core_1.ReplPlatform.JavaShell) {
            contextObject.db = this.setDbFunc(this.currentDb);
        }
        else {
            Object.defineProperty(contextObject, 'db', {
                configurable: true,
                set: setFunc,
                get: () => (this.currentDb)
            });
        }
        this.messageBus.emit('mongosh:setCtx', { method: 'setCtx', arguments: {} });
    }
    get currentServiceProvider() {
        try {
            return this.currentDb._mongo._serviceProvider;
        }
        catch (err) {
            if (err.code === error_codes_1.ShellApiErrors.NotConnected) {
                return this.initialServiceProvider;
            }
            throw err;
        }
    }
    emitApiCall(event) {
        this.messageBus.emit('mongosh:api-call', event);
    }
    emitDeprecatedApiCall(event) {
        this.messageBus.emit('mongosh:deprecated-api-call', event);
    }
    setEvaluationListener(listener) {
        this.evaluationListener = listener;
    }
    getAutocompleteParameters() {
        return {
            topology: () => {
                var _a;
                let topology;
                const topologyDescription = (_a = this.currentServiceProvider.getTopology()) === null || _a === void 0 ? void 0 : _a.description;
                const topologyType = topologyDescription === null || topologyDescription === void 0 ? void 0 : topologyDescription.type;
                switch (topologyType) {
                    case 'ReplicaSetNoPrimary':
                    case 'ReplicaSetWithPrimary':
                        topology = enums_1.Topologies.ReplSet;
                        break;
                    case 'Sharded':
                        topology = enums_1.Topologies.Sharded;
                        break;
                    default:
                        topology = enums_1.Topologies.Standalone;
                        if (topologyDescription.servers.size === 1) {
                            const [server] = topologyDescription.servers.values();
                            switch (server.type) {
                                case 'Mongos':
                                    topology = enums_1.Topologies.Sharded;
                                    break;
                                case 'PossiblePrimary':
                                case 'RSPrimary':
                                case 'RSSecondary':
                                case 'RSArbiter':
                                case 'RSOther':
                                case 'RSGhost':
                                    topology = enums_1.Topologies.ReplSet;
                                    break;
                                default:
                                    break;
                            }
                        }
                        break;
                }
                return topology;
            },
            connectionInfo: () => {
                return this.connectionInfo.extraInfo;
            },
            getCollectionCompletionsForCurrentDb: async (collName) => {
                try {
                    const collectionNames = await this.currentDb._getCollectionNamesForCompletion();
                    return collectionNames.filter((name) => name.startsWith(collName));
                }
                catch (err) {
                    if (err.code === error_codes_1.ShellApiErrors.NotConnected || err.codeName === 'Unauthorized') {
                        return [];
                    }
                    throw err;
                }
            },
            getDatabaseCompletions: async (dbName) => {
                try {
                    const dbNames = await this.currentDb._mongo._getDatabaseNamesForCompletion();
                    return dbNames.filter((name) => name.startsWith(dbName));
                }
                catch (err) {
                    if (err.code === error_codes_1.ShellApiErrors.NotConnected || err.codeName === 'Unauthorized') {
                        return [];
                    }
                    throw err;
                }
            }
        };
    }
    async onInterruptExecution() {
        this.interrupted.set();
        this.currentCursor = null;
        this.resumeMongosAfterInterrupt = await Promise.all(this.mongos.map(async (m) => {
            try {
                return {
                    mongo: m,
                    resume: await m._suspend()
                };
            }
            catch (e) {
                return {
                    mongo: m,
                    resume: null
                };
            }
        }));
        return !this.resumeMongosAfterInterrupt.find(r => r.resume === null);
    }
    async onResumeExecution() {
        var _a, _b;
        const promises = (_b = (_a = this.resumeMongosAfterInterrupt) === null || _a === void 0 ? void 0 : _a.map(async (r) => {
            if (!this.mongos.find(m => m === r.mongo)) {
                return true;
            }
            if (r.resume === null) {
                return false;
            }
            try {
                await r.resume();
                return true;
            }
            catch (e) {
                return false;
            }
        })) !== null && _b !== void 0 ? _b : [];
        this.resumeMongosAfterInterrupt = undefined;
        const result = await Promise.all(promises);
        this.interrupted.reset();
        return !result.find(r => r === false);
    }
    async getDefaultPrompt() {
        return `${this.getDefaultPromptPrefix()}${this.getTopologySpecificPrompt()}> `;
    }
    getDefaultPromptPrefix() {
        var _a, _b, _c, _d;
        const extraConnectionInfo = (_a = this.connectionInfo) === null || _a === void 0 ? void 0 : _a.extraInfo;
        if (extraConnectionInfo === null || extraConnectionInfo === void 0 ? void 0 : extraConnectionInfo.is_data_lake) {
            return 'Atlas Data Lake ';
        }
        else if ((extraConnectionInfo === null || extraConnectionInfo === void 0 ? void 0 : extraConnectionInfo.is_enterprise) || ((_d = (_c = (_b = this.connectionInfo) === null || _b === void 0 ? void 0 : _b.buildInfo) === null || _c === void 0 ? void 0 : _c.modules) === null || _d === void 0 ? void 0 : _d.indexOf('enterprise')) >= 0) {
            return 'Enterprise ';
        }
        return '';
    }
    getTopologySpecificPrompt() {
        var _a, _b, _c, _d;
        const description = (_a = this.currentServiceProvider.getTopology()) === null || _a === void 0 ? void 0 : _a.description;
        if (!description) {
            return '';
        }
        let replicaSet = description.setName;
        let serverTypePrompt = '';
        switch (description.type) {
            case 'Single':
                const singleDetails = this.getTopologySinglePrompt(description);
                replicaSet = (_b = singleDetails === null || singleDetails === void 0 ? void 0 : singleDetails.replicaSet) !== null && _b !== void 0 ? _b : replicaSet;
                serverTypePrompt = (singleDetails === null || singleDetails === void 0 ? void 0 : singleDetails.serverType) ? `[direct: ${singleDetails.serverType}]` : '';
                break;
            case 'ReplicaSetNoPrimary':
                serverTypePrompt = '[secondary]';
                break;
            case 'ReplicaSetWithPrimary':
                serverTypePrompt = '[primary]';
                break;
            case 'Sharded':
                serverTypePrompt = ((_d = (_c = this.connectionInfo) === null || _c === void 0 ? void 0 : _c.extraInfo) === null || _d === void 0 ? void 0 : _d.atlas_version) ? '' : '[mongos]';
                break;
            default:
                return '';
        }
        const setNamePrefix = replicaSet ? `${replicaSet} ` : '';
        return `${setNamePrefix}${serverTypePrompt}`;
    }
    getTopologySinglePrompt(description) {
        var _a;
        if (((_a = description.servers) === null || _a === void 0 ? void 0 : _a.size) !== 1) {
            return undefined;
        }
        const [server] = description.servers.values();
        let serverType;
        switch (server.type) {
            case 'Mongos':
                serverType = 'mongos';
                break;
            case 'RSPrimary':
                serverType = 'primary';
                break;
            case 'RSSecondary':
                serverType = 'secondary';
                break;
            case 'RSArbiter':
                serverType = 'arbiter';
                break;
            case 'RSOther':
                serverType = 'other';
                break;
            case 'Standalone':
            case 'PossiblePrimary':
            case 'RSGhost':
            case 'Unknown':
            default:
                serverType = '';
        }
        return {
            replicaSet: server.setName,
            serverType
        };
    }
}
exports.default = ShellInternalState;
