// @mongosh/service-provider-core/lib/whatwg-url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.URLSearchParams = exports.URL = exports.textEncodingPolyfill = void 0;
if (typeof require('util').TextDecoder !== 'function' ||
    typeof require('util').TextEncoder !== 'function') {
    Object.assign(require('util'), textEncodingPolyfill());
}
const _global = new Function('return this')();
let URL;
exports.URL = URL;
let URLSearchParams;
exports.URLSearchParams = URLSearchParams;
if ('URL' in _global) {
    exports.URL = URL = _global.URL;
    exports.URLSearchParams = URLSearchParams = _global.URLSearchParams;
}
else {
    exports.URL = URL = require('whatwg-url').URL;
    exports.URLSearchParams = URLSearchParams = require('whatwg-url').URLSearchParams;
}
function textEncodingPolyfill() {
    class TextEncoder {
        encode(string) {
            return Buffer.from(string, 'utf8');
        }
    }
    class TextDecoder {
        decode(bytes) {
            const str = Buffer.from(bytes).toString('utf8');
            return str.slice(+str.startsWith('\ufeff'));
        }
    }
    return { TextDecoder, TextEncoder };
}
exports.textEncodingPolyfill = textEncodingPolyfill;
